### PR TITLE
Constraint files are now maintained automatically

### PR DESCRIPTION
**Pull request by potiuk**
* No need to generate requirements when setup.py changes
* requirements are kept in separate orphan branches not in main repo
* merges to master verify if latest requirements are working and
  push tested requirements to orphaned branches
* we keep history of requirement changes and can label them
  individually for each version (by constraint-1.10.n tag name)
* consistently changed all references to be 'constraints' not
  'requirements'

---
Make sure to mark the boxes below before creating PR: [x]

- [x] Description above provides context of the change
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [x] Target Github ISSUE in description if exists
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [x] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

### CODE REVIEW

**Comment by potiuk** [comment]
Hello. (also cc @ephraimbuddy who experienced problems with the previous requirement generation). 

I think I found a very good and workable solution for former "requirements" (now "constraints"). 

In short:

- I keep constraints*.txt files in separate (orphaned) branches - similar to what github-pages does. This way "constraints" can evolve separately from the main code (for example constraint changes in master are kept in the `constraints-master` branch). Similarly 1.10 requirements will be kept in `constraints-1-10` branch.

- The new "upgraded" constraints are pushed automatically by merge to master (or push to v1-10-test) - such pushes will always try to upgrade to latest constraints and in case tests pass, they will push a new commit to the  `constraints-master` (or `constraint-1-10` branch. The commit will include run id and commit hash of the original commit from master/v1-10-test - so that we know which commit brought the new constraints in. This will only happen if all the tests are successful, so there is no risk we push any constraint change to that branch if any of the tests fail.

- @kaxil  @dimberman  - for releases we will just have to tag the constraints with 'constraints-1.10.12` (for example) and this way our Dockerfiles and `pip install` will be able to use those "fixed" set of requirements  

- all the other properties of constraints/requirements are kept - we have open-ended setup.py requirements, but also a way to manage (fully automatically) a set of good, proven-by-tests constraint.txt files that can be used for repeatable/reproducible  airflow installation and docker image generation. 

- last but not least: I used the opportunity to replace 'requirements' -> 'constraints' in all related places in docs, file names etc. - the "requirements" name was extremely confusing.

**Comment by ephraimbuddy** [comment]
This is great! I spent a lot of time generating requirements the other day. Thanks for this 😊

File: BREEZE.rst [hunk 1]
```diff
@@ -693,31 +693,37 @@ easily identify the location the problems with documentation originated from.
...
+  ./breeze generate-constraints --python 3.8
 
+This bumps the constraint files to latest versions and stores hash of setup.py. The generated constraint
+and setup.py hash files are stored in the ``files`` folder and while generating the constraints diff
+of changes vs the previous constraint files is printed.
 
-This bumps requirements to latest versions and stores hash of setup.py so that we are automatically
-upgrading the requirements as we add new ones.
+Note - the movie below shows previous version of the command ``generate-requirements`` that is not valid
+any-more.
```
Comment (turbaszek) [inline BREEZE.rst]: Should we keep it then?

File: CONTRIBUTING.rst [hunk 2]
```diff
@@ -342,76 +342,61 @@ be open to allow several different libraries with the same requirements to be in
...
+automatically and committed to orphan ``constraints-master`` and ``constraint-1-10`` branches based on
+the set of all latest working and tested dependency versions. Those
+``constraints-<PYTHON_MAJOR_MINOR_VERSION>.txt`` files can be used as
+constraints file when installing Apache Airflow - either from the sources:
 
 .. code-block:: bash
 
-  pip install -e . --constraint requirements/requirements-python3.6.txt
+  pip install -e . \
+    --constraint --constraint "https://raw.githubusercontent.com/apache/airflow/constraints-master/constraints-3.6.txt"
```
Comment (turbaszek) [inline CONTRIBUTING.rst]:
```suggestion
    --constraint "https://raw.githubusercontent.com/apache/airflow/constraints-master/constraints-3.6.txt"
```

File: CONTRIBUTING.rst [hunk 3]
```diff
@@ -342,76 +342,61 @@ be open to allow several different libraries with the same requirements to be in
...
 
 
-or from the pypi package
+or from the pypi package:
 
 .. code-block:: bash
 
-  pip install apache-airflow --constraint requirements/requirements-python3.6.txt
+  pip install apache-airflow \
+    --constraint --constraint "https://raw.githubusercontent.com/apache/airflow/constraints-master/constraints-3.6.txt"
```
Comment (turbaszek) [inline CONTRIBUTING.rst]:
```suggestion
    --constraint "https://raw.githubusercontent.com/apache/airflow/constraints-master/constraints-3.6.txt"
```

File: CONTRIBUTING.rst [hunk 4]
```diff
@@ -342,76 +342,61 @@ be open to allow several different libraries with the same requirements to be in
...
 
 
 This works also with extras - for example:
 
 .. code-block:: bash
 
-  pip install .[ssh] --constraint requirements/requirements-python3.6.txt
+  pip install .[ssh] \
+        --constraint --constraint "https://raw.githubusercontent.com/apache/airflow/constraints-master/constraints-3.6.txt"
```
Comment (turbaszek) [inline CONTRIBUTING.rst]:
```suggestion
        --constraint "https://raw.githubusercontent.com/apache/airflow/constraints-master/constraints-3.6.txt"
```

File: IMAGES.rst [hunk 5]
```diff
@@ -92,14 +92,14 @@ parameter to Breeze:
...
+  ./breeze build-image --python 3.7 --extras=gcp --production-image --install-airflow-version=1.10.12
 
 This will build the image using command similar to:
 
 .. code-block:: bash
 
-    pip install apache-airflow[sendgrid]==1.10.9 \
-       --constraint https://raw.githubusercontent.com/apache/airflow/v1-10-test/requirements/requirements-python3.7.txt
+    pip install apache-airflow[sendgrid]==1.10.12 \
+  --constraint --constraint "https://raw.githubusercontent.com/apache/airflow/constraints-1.10.12/constraints-3.6.txt"
```
Comment (turbaszek) [inline IMAGES.rst]:
```suggestion
  --constraint "https://raw.githubusercontent.com/apache/airflow/constraints-1.10.12/constraints-3.6.txt"
```

File: IMAGES.rst [hunk 6]
```diff
@@ -109,8 +109,8 @@ parameter to Breeze:
 
 .. code-block:: bash
 
-    pip install https://github.com/apache/airflow/archive/<tag>.tar.gz#egg=apache-airflow \
-       --constraint https://raw.githubusercontent.com/apache/airflow/<tag>/requirements/requirements-python3.7.txt
+    pip install "https://github.com/apache/airflow/archive/<tag>.tar.gz#egg=apache-airflow" \
+  --constraint --constraint "https://raw.githubusercontent.com/apache/airflow/constraints-master/constraints-3.6.txt"
```
Comment (turbaszek) [inline IMAGES.rst]:
```suggestion
  --constraint "https://raw.githubusercontent.com/apache/airflow/constraints-master/constraints-3.6.txt"
```

**Review by turbaszek — COMMENTED** [review]
I love the change because I've already had to generate those files a few times. At first, I was concerned about the naming but I see that `pip install` has the `--constraint` flag so now I understand.

File: BREEZE.rst (hunk 1 above)
Comment (potiuk) [reply BREEZE.rst]: Good question. probably not.

Comment (potiuk) [reply BREEZE.rst]: REmoved

File: CONTRIBUTING.rst (hunk 2 above)
Comment (potiuk) [reply CONTRIBUTING.rst]: aaargh. Resolved all of them,

File: docs/installation.rst [hunk 7]
```diff
@@ -52,19 +53,17 @@ and python versions in the URL.
 .. code-block:: bash
 
     pip install \
-     apache-airflow==1.10.10 \
-     --constraint \
-            https://raw.githubusercontent.com/apache/airflow/1.10.10/requirements/requirements-python3.7.txt
+     apache-airflow==1.10.12 \
+     --constraint "https://raw.githubusercontent.com/apache/airflow/constraints-1.10.12/constraints-3.7.txt"
```
Comment (kaxil) [inline docs/installation.rst:57]: This is a broken link.

Correct link
```
https://raw.githubusercontent.com/apache/airflow/constraints-1-10/constraints-3.7.txt
```
But does that mean 1.10.10 and 1.10.11 have the some constraint file?

Comment (potiuk) [reply docs/installation.rst:57]: It won't be broken when we release 1.10.12 and tag it.

Good point with 1.10.10 == 1.10.11  @kaxil . I thought they will work anyway, but we can do better easily.

Currently it is the same I just added one commit in the orphaned branch  and I added all tags there. But there is nothing to take the "frozen" requirements from 1.10.10 and add extra commit and tag it. I will do it momentariily.

**Review by turbaszek — APPROVED** [review]

File: docs/installation.rst (hunk 7 above)
Comment (potiuk) [reply docs/installation.rst:57]: Now - I recovered 1.10.10 and 1.10.11 and they are separate commits : 

```
commit f53b1d0041d3dc087ff2c6addaa72266e784b928 (HEAD -> constraints-1-10, tag: constraints-1.10.11, apache/constraints-1-10, apache-ssh/constraints-1-10)
Author: Jarek Potiuk <jarek@potiuk.com>
Date:   Mon Jul 20 13:14:27 2020 +0200

    Added constraints for 1.10.11

commit 1c2d5de877a993a6541e017d4d59938315814448 (tag: constraints-1.10.9, tag: constraints-1.10.8, tag: constraints-1.10.7, tag: constraints-1.10.6, tag: constraints-1.10.5, tag: constraints-1.10.4, tag: constraints-1.10.3, tag: constraints-1.10.2, tag: constraints-1.10.10, tag: constraints-1.10.1)
Author: Jarek Potiuk <jarek@potiuk.com>
Date:   Sun Jul 19 19:35:22 2020 +0200

    Add orphan constraints for 1-10 branch 1.10.10 version


```

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -443,6 +412,97 @@ ${{ hashFiles('requirements/requirements-python${{matrix.python-version}}.txt')
...
+      fail-fast: false
+    needs:
+      - cancel-previous-workflow-run
+      - tests-sqlite
+      - tests-mysql
+      - tests-postgres
+      - tests-kubernetes
+    env:
+      PYTHON_MAJOR_MINOR_VERSION: ${{ matrix.python-version }}
+      CI_JOB_TYPE: "Requirements"
```
Comment (kaxil) [inline .github/workflows/ci.yml]: Does this need to change?  (I am not sure --- just asking :) )

i.e. we have changed `requirements` to `constraints` elsewhere so just wondering

**Review by kaxil — APPROVED** [review]

File: .github/workflows/ci.yml (hunk 8 above)
Comment (potiuk) [reply .github/workflows/ci.yml]: aaaargh.. Missed that one :). In fact we do not need CI_JOB_TYPE at all any more (it was used in Travis). So I just deleted all of them now :)

**Comment by potiuk** [comment]
Nice! First automated commit from @ephraimbuddy's fix to limit the google client :). @ephraimbuddy I think that was a much nicer experience :) https://github.com/apache/airflow/commit/99fa5ef4bb8f69d76e8576c347cea042f4b00ada

**Comment by ephraimbuddy** [comment]
Sure. Very nice experience. Previously, it would take me hours to generate requirements for the 3 python version, but yesterday, I did the change  and pushed under 10mins. Thanks so much for this 🙌

**Pull request merged**